### ISSUE TITLE
Analysis: Why chat templates reduce DeepSeek MTP speculative decoding accept rate

### ISSUE THREAD

**Issue by carlushuang** [ROCm/ATOM#609]
## Summary

This is a general analysis of why enabling `--use-chat-template` (or equivalent chat formatting) reduces the acceptance rate of DeepSeek's Multi-Token Prediction (MTP) module when used for speculative decoding. This affects DeepSeek-V3, V3.1, R1, and derivatives.

## Root Cause: Distribution Mismatch Between Pretraining and Post-Training

DeepSeek's training pipeline is:

```
Pretraining (14.8T tokens, raw text) → SFT (chat finetuning) → RLHF
         ↑                                    ↑
    MTP module trained HERE            MTP module NOT updated HERE
```

The MTP module (14B params) was trained jointly with the main model (671B) during **pretraining** on raw web/text data. When the main model subsequently went through SFT and RLHF to become a chat model, the **MTP module's weights were frozen** — it was not updated during post-training.

This means:
- The **main model's output distribution shifted** from "predict next token in raw text" to "predict next token in chat-formatted, instruction-following, RLHF-aligned text"
- The **MTP draft model still reflects the pretraining distribution**

When `--use-chat-template` is enabled, the input is forced into chat format (`<|begin_of_sentence|><|User|>...<|Assistant|>...`), which is the distribution the main model was fine-tuned for but the MTP module was **not**. Without the chat template (raw prompt), the input looks more like pretraining data, which is closer to what the MTP module expects.

### 1. Special Tokens Are High-Entropy Decision Points

Chat templates inject special tokens (`<|User|>`, `<|Assistant|>`, `<|end_of_sentence|>`, thinking tags for R1, etc.) that create **structural transition points** in the token sequence. At these boundaries:

- The main model has learned (via SFT/RLHF) exactly which special token comes next
- The MTP module, trained only on pretraining data, has weak or no signal for these tokens
- Result: the draft token is almost always **rejected** at structural boundaries

This compounds because a single rejection in speculative decoding throws away all subsequent draft tokens in that batch — one bad prediction at a role boundary wastes the entire speculation window.

### 2. RLHF Shifts the Token Distribution Even Within Natural Text

Even ignoring special tokens, RLHF changes *how* the model writes. Chat-tuned models produce more structured, cautious, hedging language that differs from raw pretraining text. The MTP module's predictions diverge from the main model even on "normal" tokens because the stylistic distribution has shifted.

### 3. Autoregressive Reuse Degrades Later Positions

The MTP module was trained for single next-token prediction but is reused autoregressively across multiple draft positions. This compounds the distribution mismatch at each successive position, leading to rapidly declining accept rates for k > 1.

## Summary Table

| Factor | Without chat template | With chat template |
|--------|----------------------|-------------------|
| Input distribution | Closer to pretraining data | Chat-formatted (SFT/RLHF distribution) |
| Special tokens | Few/none | Many role markers, thinking tags |
| MTP module calibration | Better match | Distribution mismatch |
| Accept rate | Higher | Lower |

## Evidence

- **Nebius** ([MTP-DeepSeek-V3-0324](https://huggingface.co/nebius/MTP-DeepSeek-V3-0324)) explicitly identified this mismatch and fine-tuned the MTP module with "LK losses" that directly optimize acceptance rate rather than using KL divergence as a proxy.
- **Perplexity** ([blog](https://www.perplexity.ai/hub/blog/accelerating-sonar-through-speculation)) trains custom MTP heads on their fine-tuned models' actual output distribution, rather than relying on DeepSeek's pretraining-era MTP weights.
- **vLLM benchmarks** ([#27726](https://github.com/vllm-project/vllm/issues/27726)) show acceptance rates of ~83% for k=1 on ShareGPT (chat data), below the ~90% reported in the DeepSeek paper (likely measured on pretraining-like data).
- **arXiv 2604.14682** ([Acceptance Dynamics Across Cognitive Domains](https://arxiv.org/abs/2604.14682)) shows that task type is a stronger predictor of acceptance than tree depth, confirming that distribution characteristics dominate accept rate.

## Possible Mitigations

1. **Fine-tune MTP module on chat data** — as Nebius and Perplexity have done, retrain the MTP weights to match the post-trained model's distribution
2. **Use LK losses** instead of KL divergence when fine-tuning, to directly optimize acceptance rate
3. **Reduce speculation depth** when using chat templates (k=1 instead of k=2+) to limit the compounding mismatch
4. **Adaptive speculation** — dynamically reduce draft length near structural boundaries (role tokens, thinking tags)

## References

- [DeepSeek-V3 Technical Report (arXiv 2412.19437)](https://arxiv.org/html/2412.19437v1)
- [Nebius MTP-DeepSeek-V3-0324](https://huggingface.co/nebius/MTP-DeepSeek-V3-0324)
- [Perplexity: Accelerating Sonar Through Speculation](https://www.perplexity.ai/hub/blog/accelerating-sonar-through-speculation)
- [vLLM MTP Issue #14137](https://github.com/vllm-project/vllm/issues/14137)
- [SGLang DeepSeek MTP Docs](https://docs.sglang.ai/basic_usage/deepseek_v3.html)
- [ROCm MTP Tutorial](https://rocm.docs.amd.com/projects/ai-developer-hub/en/latest/notebooks/inference/mtp.html)

**Comment by maodoudou168** [comment]
Hi, the analysis is impressive! I think have experienced similar situations but with a bit difference, I was trying to retrain GLM's MTP based on some online data (about 10k online coding data with chat-template), however, when doing tests, the accepet length on samples without chat-template could increase around 7%, but when the chat-template was applied, the accept length increment compared with the original MTP collapsed to around 2.6%. I haven't found a good method to solve this problem yet, after reading your blog, I've tried LK loss but it showed no improvements. I was wondering if I could mix more standard chat-template data into the online data during training or just increasing the training data size would tackle this issue. May I know if you have any suggestions on this? Thanks a lot.
